Power optimization: Scan screen only when user touches it

sarav opened the pull request:
This addresses #300 

The current implementation scans the screen every 750 milliseconds. This
is pretty expensive in terms of power consumption.

Since the screen mode (map, gym, pokemon, etc) doesn't change unless the
user actually touches the screen, we just need to scan the screen around
the time the screen is touched.

Also, since the user touches the screen a lot during a gym battle, add
some throttling to scan the screen only after the last touch if multiple
touches happen in quick succession (within 750 milliseconds).

This should significantly improve the power.

Since there are some animations which take time when the user opens a
pokemon switches pokemon, scan a couple of times before stopping.

**Caveat:** The touch events only come when the user touches the screen and not when the lift the finger. So, if they do a long drag (2.5 seconds), the scanning would finish before they lift the finger. But this hasn't really happened to me when I use the game UI normally. Also, in that case, they just need to touch the screen once to trigger the scan again. So, it's not that bad for the enormous power savings.

**Result:** Not sure if it's placebo, but I feel like the device is running cooler. Also, this is the first time I was able to start GoIV and then go back to the game that was already running without having to restart it.

Blaisorblade commented on app/src/main/java/com/kamron/pogoiv/Pokefly.java:
```diff
@@ -364,26 +368,59 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+                if (screenScanRetries > 0) {
+                    scanPokemonScreen();
+                    screenScanRetries--;
+                    screenScanHandler.postDelayed(screenScanRunnable, 500);
```
I hadn't realized we could have used `postDelayed` instead of `TimerTask` *all this time* 😢

Blaisorblade replied:
This is the 500 I'd replace by `SCREEN_SCAN_RETRY_DELAY_MS`.

sarav replied:
This was supposed to be 750 ms too. So, just used the same final variable.

Blaisorblade commented on app/src/main/java/com/kamron/pogoiv/Pokefly.java:
```diff
@@ -364,26 +368,59 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+            public boolean onTouch(View v, MotionEvent event) {
+                if (event.getActionMasked() == MotionEvent.ACTION_OUTSIDE) {
+                    screenScanHandler.removeCallbacks(screenScanRunnable);
+                    screenScanHandler.postDelayed(screenScanRunnable, 750);
```
Could use `screenScreenDelay` rather than `750`.

Blaisorblade commented on app/src/main/java/com/kamron/pogoiv/Pokefly.java:
```diff
@@ -126,6 +123,13 @@
+    private WindowManager.LayoutParams touchViewParams;
+    private Handler screenScanHandler;
+    private Runnable screenScanRunnable;
+    private final int screenScreenDelay = 750;
```
I'd make it static and add the unit of measure, so `screenScreenDelayMs`.

sarav replied:
This was the whole point of adding this variable, but I forgot to use it. Did you test it? How well did it work for you?

Blaisorblade replied:
Yeah I guessed, `screenScreen` was also telling ;-). Testing now...

sarav replied:
Working late at night doesn't help :)

Blaisorblade commented:
This looks good. I've added a couple of comments. And overall I might prefer having this in a separate small class since (1) we can (2) this would encapsulate a small state machine (3) Pokefly is too big, so this small state machine is lost amid a bunch of others.

But I won't insist on that too much since it is in part a matter of preference.

Blaisorblade commented on app/src/main/java/com/kamron/pogoiv/Pokefly.java:
```diff
@@ -126,6 +123,13 @@
+    private LinearLayout touchView;
+    private WindowManager.LayoutParams touchViewParams;
+    private Handler screenScanHandler;
+    private Runnable screenScanRunnable;
```
The code would be easier to navigate if this were final or we used a non-anonymous subclass.

sarav replied:
Why final? I want to release them when the service isn't running or not allocate them when we are using battery saver mode.

sarav replied:
I'm choosing to keep it this way for the above reason. Let me know if you have a strong opinion/reason for changing it.

Blaisorblade replied:
I only meant final for the `screenScanRunnable`—it's just one instance. But it's a minor thing either way.

sarav commented:
@Blaisorblade regarding your comment about moving this code to a different class, that'll unfortunately force us to add an intent to take the screenshot. I agree in principle, but I think we should decide this after seeing how Pokefly looks after we move the dialog box UI code out -- there's SO MUCH of it!

I almost want to say we shouldn't allow any UI changes until these are addressed.

Blaisorblade commented:
> I almost want to say we shouldn't allow any UI changes until these are addressed.

That was the plan, but last we tried there was nobody doing any work (because as you told me you were on vacation). If you have the time to try we can schedule that, keeping into account open PRs.

Blaisorblade commented:
I've pushed a few extra commits on `optimize-screen-grab-mod` (haven't removed the logging yet), that's enough to fix it for me. I hope that works well enough.

Blaisorblade commented:
BTW we should maybe use this: https://github.com/blog/2247-improving-collaboration-with-forks

sarav commented:
I took a quick look. Yeah, I had already implemented and threw away the code to stop scan it the button status changed. Because there are cases like when you swipe between Pokemon, where if you do it slowly enough, the button can disappear and appear again. 

So, I think all you need is an increase in retry count or the delay (looks like it just takes a while for the UI to settle for you).

Are you okay with just increasing the retry by 1? It's still way more efficient that what we do today.

sarav commented:
I'm more curious about fixing the null pointer exception even if it didn't cause an issue for you.

Also understood you weren't suggestion a new activity, but I think we'd still have to use intents nonetheless to avoid abstraction breaking hacks. I'll look into it later as a separate refactoring improvement. As is, I don't think I made it any worse.

sarav commented:
I need to fix some typos, but on mobile now. Will fix later in about 5 hrs.

Blaisorblade commented:
> Are you okay with just increasing the retry by 1? It's still way more efficient that what we do today.

With 3 it still had issues sometimes (when loading the first pokemon), using 4 (or a bigger delay) should be better.
I'm mainly afraid of discoverability and I'm sure some users will think it broke if it doesn't work. Even if we say it first thing in the changelog and announcement.

> Also understood you weren't suggestion a new activity, but I think we'd still have to use intents nonetheless to avoid abstraction breaking hacks. I'll look into it later as a separate refactoring improvement. As is, I don't think I made it any worse.

Can be later. I'd give the class a pointer to Pokefly. Can try myself.

Blaisorblade commented on app/src/main/java/com/kamron/pogoiv/Pokefly.java:
```diff
@@ -364,26 +368,59 @@ public int onStartCommand(Intent intent, int flags, int startId) {
+                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
+                        | WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
+                        | WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH,
+                PixelFormat.TRANSLUCENT);
```
Can we get by with TRANSPARENT? We need to be fully transparent anyway...
```
    /** System chooses a format that supports translucency (many alpha bits) */
    public static final int TRANSLUCENT = -3;

    /**
     * System chooses a format that supports transparency
     * (at least 1 alpha bit)
     */
    public static final int TRANSPARENT = -2;
```

sarav commented:
Ok, I'll increase retires and change to transparent and test it. And I'll push it.

I might later try increasing the delay between retries (and reducing the no. of retries left) if the button state changes.

sarav commented:
**Warning:** Rambling comment.

> Can be later. I'd give the class a pointer to Pokefly.

Giving a pointer to Pokefly and making the methods public is what I meant by "abstraction breaking hacks". If it needs to be that tightly coupled, then I'm not sure there's much benefit to moving it out of Pokefly. Pure logic, like OCR, scanning the screen, etc I can see why it's good to move it out. Also, the main point of Pokefly is to stay in background, check for the right conditions and throw up the IV button -- so this really should stay in Pokefly.

If anything (as in move anything other than UI code out of Pokefly), we could/should move the screen identifying code into a pure logic class that'll identify the screen we are on (Map, pokemon, gym, etc). Say, ScreenIdentifier.java. We'd initialize it with the screen dimensions, a screengrabber object (to getPixels from), etc and then later do stuff like screenIdentifier.isPokeScreen().

If I were to ever move the screen watching code out of Pokefly in a clean way. I'd want to move this into a separate class and make it behave like ScreenShotHelper. Similar to screenshot helper looking at some events to go find the pokemon bitmap (screenshot) to scan, this new class would look at the touch events, check if the screen has a pokemon and if so, send the process-bitmap intent. But at that point, it'll be more of the case of not showing an IV button and directly calculating the IV and displaying a floating IV result that someone can touch to expand. That way, it'd work just as well for screenshot or automated screen scans. At that point, there's really nothing to do in Pokefly (since we won't be showing the IV button either).

sarav commented:
> I took a quick look. Yeah, I had already implemented and threw away the code to stop scan it the button status changed. Because there are cases like when you swipe between Pokemon, where if you do it slowly enough, the button can disappear and appear again.

@Blaisorblade, took another look at your code. My statement about shouldn't be an issue. The code I added was to stop scanning if the button went from "shown" to "not shown" or "not shown" to "shown". But you are stopping only when showing the button. Since there's no way make the IV button appear and then exit from the Pokemon screen in one touch/swipe event, I think your change should be fine. I pulled in your commits into my branch, cleaned up the logs and trying them out. If it works fine, I'll update this PR to include your changes too.

sarav commented:
@Blaisorblade Travis seems to have thrown up in his mouth. Can you revive him?

Blaisorblade commented:
Merged #410, confirmed that's the issue. To fix Travis please rebase on latest master.
